src/tactic/core.lean

```diff
@@ -64,14 +66,32 @@ end expr
 namespace interaction_monad
 open result
 
+variables {σ : Type} {α : Type u}
+
+/-- `get_state` returns the underlying state inside an interaction monad, from within that monad. -/
+meta def get_state : interaction_monad σ σ :=
+λ state, success state state
+
+/-- `set_state` sets the underlying state inside an interaction monad, from within that monad. -/
+meta def set_state (state : σ) : interaction_monad σ unit :=
+λ _, success () state
+
+/-- `under_state` applies the given state to `tac`, and returns the result. If `tac` fails, then
+`under_state` does too. -/
+meta def under_state (state : σ) (tac : interaction_monad σ α) : interaction_monad σ α :=
```

Comment:
Maybe name as `run_with_state`? `under_state` is not so obvious to me.

Reply:
Also, what is the final state? If the final state is the same as the initial state, maybe something like `with_local_state` would convey that idea better.

Reply:
Just because `local` has lean-specific technical connotations, I'd prefer to avoid it(?). I can't think of a great name, so unless someone objects I'll copy Scott's.